modules/openapi-generator/src/main/java/org/openapitools/codegen/OpenAPINormalizer.java

```diff
@@ -0,0 +1,427 @@
+        if (requestBody == null) {
+            return;
+        }
+
+        // unalias $ref
+        if (requestBody.get$ref() != null) {
+            String ref = ModelUtils.getSimpleRef(requestBody.get$ref());
+            requestBody = openAPI.getComponents().getRequestBodies().get(ref);
+
+            if (requestBody == null) {
+                return;
+            }
+        }
+
+        normalizeContent(requestBody.getContent());
+    }
+
+    /**
+     * Normalizes schemas in parameters
+     *
+     * @param operation target operation
+     */
+    private void normalizeParameters(Operation operation) {
+        List<Parameter> parameters = operation.getParameters();
+        if (parameters == null) {
+            return;
+        }
+
+        for (Parameter parameter : parameters) {
+            if (parameter.getSchema() == null) {
+                continue;
+            } else {
+                normalizeSchema(parameter.getSchema(), new HashSet<>());
+            }
+        }
+    }
+
+    /**
+     * Normalizes schemas in ApiResponses
+     *
+     * @param operation target operation
+     */
+    private void normalizeResponses(Operation operation) {
+        ApiResponses responses = operation.getResponses();
+        if (responses == null) {
+            return;
+        }
+
+        for (Map.Entry<String, ApiResponse> responsesEntry : responses.entrySet()) {
+            if (responsesEntry.getValue() == null) {
+                continue;
+            } else {
+                normalizeContent(responsesEntry.getValue().getContent());
+            }
+        }
+    }
+
+    /**
+     * Normalizes schemas in components
+     */
+    private void normalizeComponents() {
+        Map<String, Schema> schemas = openAPI.getComponents().getSchemas();
+        if (schemas == null) {
+            return;
+        }
+
+        List<String> schemaNames = new ArrayList<String>(schemas.keySet());
+        for (String schemaName : schemaNames) {
+            Schema schema = schemas.get(schemaName);
+            if (schema == null) {
+                LOGGER.warn("{} not fount found in openapi/components/schemas.", schemaName);
+            } else {
+                normalizeSchema(schema, new HashSet<>());
+            }
+        }
+    }
+
+    /**
+     * Normalizes a schema
+     *
+     * @param schema Schema
+     * @param visitedSchemas a set of visited schemas
+     */
+    public void normalizeSchema(Schema schema, Set<Schema> visitedSchemas) {
+        if (schema == null) {
+            return;
+        }
+
+        if (StringUtils.isNotEmpty(schema.get$ref())) {
+            // not need to process $ref
+            return;
+        }
+
+        if ((visitedSchemas.contains(schema))) {
+            return; // skip due to circular reference
+        } else {
+            visitedSchemas.add(schema);
+        }
+
+        if (schema instanceof ArraySchema) {
+            if (schema.getItems() != null) {
```

Comment:
🤔 I would extract every block of the enclosing `if-else` into its own method to reduce the complexity/length of this method.